Seed storage modules

rihardsgravis opened the pull request:
# Description

Decouple seed storing, address generation and signing from state actions to separate modules.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

Tested on macOS, Ubuntu, iOS.

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] For changes to `mobile` that include native code (including React Native modules): I have verified that both iOS and Android successfully build in both `Debug` and `Release` modes
- [x] For changes to `shared`: If applicable, I have verified that my changes are implemented correctly in `desktop` and `mobile`

Commit:
Remove redundant `accounts.seedCount` state

Commit:
Remove redundant `accounts.accountNames` state

Commit:
Initial Vault module implementation

Commit:
Merge branch 'develop' into feat/seed-storage-modules

* develop: (22 commits)
  Make sure __tests__ mimic src/ directory structure
  Rename iota-wallet-shared-modules to shared-modules
  Use babel-plugin-module-resolver in mobile directory
  Reorganise mobile directory structure
  Mobile/iOS: Use NodeMobile framework built with Bitcode
  Mobile: Fix paths, update post install
  Mobile: Update rn-cli-config
  Mobile: Fix wallet reset bug
  Mobile: Reorder save seed options
  Mobile: Adjust checksum position
  Mobile/iOS: Re-enable bitcode
  New Crowdin translations [ci skip] (#354)
  Mobile: Fix Android account name cut-off
  Mobile: Fix chart interpolation Android
  Mobile: Fix Android keyboard avoidance
  Mobile: Remove react-native-content-aware-scroll-view
  Trigger 2fa verification on Settings view automatically once necessary code length is reached.
  Clear 2fa code value after successful verification
  Add code documentation
  Trigger 2fa verification once necessary code string length is reached
  ...

# Conflicts:
#	src/mobile/src/ui/components/TopBar.js
#	src/mobile/src/ui/views/wallet/EditAccountName.js
#	src/mobile/src/ui/views/wallet/UseExistingSeed.js

Commit:
Merge branch 'develop' into feat/seed-storage-modules

* develop:
  Fix SafetyNet script [ci skip]
  Reattach only if transaction falls below max depth
  Update changelog
  Promotion adjustments (#368)
  Remove all references to iota-wallet-shared-modules (#369)
  Shared: Fix available balance selector
  New translations translation.json (Swedish)
  New translations translation.json (Swedish)
  New Crowdin translations [ci skip] (#363)
  Mobile: Bump build numbers for release v0.5.0 (22)

Commit:
Update mobile to seed storage modules

Commit:
Merge branch 'develop' into feat/seed-storage-modules

* develop:
  Use `byteTochar` at all occurrences.

# Conflicts:
#	src/desktop/src/ui/views/wallet/Receive.js

cvarley100 commented on src/desktop/src/libs/vault/Keychain.js:
```diff
@@ -0,0 +1,213 @@
+     * @param {array} seed - Byte array seed
+     * @returns {promise} - Resolves to a success boolean
+     */
+    accountAdd = async (accountId, seed) => {
```
Maybe addAccount, removeAccount and renameAccount?

cvarley100 commented on src/desktop/src/ui/Index.js:
```diff
@@ -179,7 +178,7 @@ class App extends React.Component {
      */
     accountSwitch(accountName) {
-        const accountIndex = this.props.accounts.accountNames.indexOf(accountName);
+        const accountIndex = Object.keys(this.props.accounts.accountInfo).indexOf(accountName);
```
This could be a selector

cvarley100 commented on src/mobile/src/ui/views/onboarding/SetAccountName.js:
```diff
@@ -108,64 +107,56 @@ export class SetAccountName extends Component {
             );
         }
+
+        if (accountNames.map((item) => item.toLowerCase()).indexOf(accountName.toLowerCase()) > -1) {
```
Maybe add this to SeedStore as `isAccountNameUnique` as it'll be used when adding accounts, changing account name etc.

rihardsgravis replied:
A single `seedStore` contains account info only about it's seedStore type (on desktop they are also hashed), so it would require to loop all account seedStore types to do this check.
The state should always contain the correct account name list.

cvarley100 commented on src/mobile/src/ui/views/wallet/UseExistingSeed.js:
```diff
@@ -173,19 +171,19 @@ class UseExistingSeed extends Component {
      */
     getDefaultAccountName() {
         const { t } = this.props;
-        if (this.props.seedCount === 0) {
```
Probably just keep the name suggestion empty

cvarley100 commented on src/mobile/src/ui/views/wallet/UseExistingSeed.js:
```diff
@@ -261,28 +263,19 @@ class UseExistingSeed extends Component {
-                .catch((err) => console.log(err)); // eslint-disable-line no-console
+
+            const vault = new Vault.keychain(password);
+            const isUniqeSeed = vault.uniqueSeed(password, seed);
```
isUniqueSeed*

cvarley100 commented on src/mobile/src/ui/views/wallet/UseExistingSeed.js:
```diff
@@ -261,28 +263,19 @@ class UseExistingSeed extends Component {
-                    if (isNull(seedInfo)) {
-                        return this.fetchAccountInfo(seed, accountName);
-                    }
-                    if (hasDuplicateAccountName(seedInfo, accountName)) {
```
How come we aren't checking for unique acc name?

rihardsgravis replied:
The check is happening at line https://github.com/iotaledger/trinity-wallet/blob/feat/seed-storage-modules/src/mobile/src/ui/views/wallet/UseExistingSeed.js#L256

Commit:
Bugfixes and updates:

- Move vault clear to libs/crypto from SeedVault
- Rename Vault to SeedStore
- Rename SeedStore functions to addAccount, removeAccount and renameAccount, isUniqueSeed
- Update getAccountNamesFromState selector use
- Update mobile unique account name check for additional seed

rajivshah3 commented on src/desktop/src/ui/views/settings/Advanced.js:
```diff
@@ -59,11 +59,11 @@ class Advanced extends PureComponent {
         const { t, generateAlert } = this.props;
 
         try {
-            await clearVault(password);
```
Since the `await` was removed here I think the `async` is no longer needed

Commit:
Remove unnecessary async

Commit:
Remove default account name suggestions

Commit:
Fix `dispatchErrors` not triggered on desktop

Commit:
Bugfixes and updates:

- Clear keychain on init
- Wait for `setTwoFa` to execute
- Fix `accountType` typo
- Set seed on View seed view correctly

Commit:
Check on keychain availability on keychain initialisation. Fixes #260

Commit:
Fix `setAdditionalAccountInfo` use on SeedVault import

cvarley100 commented on src/mobile/src/ui/views/onboarding/SetPassword.js:
```diff
@@ -111,9 +111,9 @@ class SetPassword extends Component {
+        const seedStore = new SeedStore.keychain(pwdHash);
 
-        const isUniqueueSeed = vault.uniqueSeed(seed);
+        const isUniqueueSeed = seedStore.isUniqueSeed(seed);
```
isUnique* :trollface:

cvarley100 commented on src/desktop/src/ui/views/onboarding/AccountName.js:
```diff
@@ -35,9 +35,7 @@ class AccountName extends React.PureComponent {
-        name: this.props.wallet.additionalAccountName.length
-            ? this.props.wallet.additionalAccountName
-            : Object.keys(this.props.accountNames).length === 0 ? this.props.t('mainWallet') : '',
+        name: this.props.wallet.additionalAccountName.length ? this.props.wallet.additionalAccountName : '',
```
Ah I would keep the first suggestion as `MAIN WALLET` but leave it blank from thereon in

cvarley100 commented on src/desktop/src/ui/views/onboarding/AccountPassword.js:
```diff
@@ -79,7 +79,11 @@ class AccountPassword extends React.PureComponent {
-        await initKeychain();
+        try {
+            await initKeychain();
+        } catch (e) {
```
👍

Commit:
Merge branch 'develop' into feat/seed-storage-modules

* develop:
  Version bump and Changelog update (#408)
  Fix trit/byte array conversion
  Accept key 9 as a valid input
  Remove transaction `hash` key reference and update search filter conditions
  New Crowdin translations [ci skip] (#373)
  Bump shared `iota.lib.js` dependency version
  Update Changelog
  Version bump and Changelog update
  Dependencies update and reverse `iota.lib.js` use from shared modules
  Code review fixes
  [Snyk] Fix for 3 vulnerable dependencies (#397)
  Fix typos and comment code
  [Snyk] Fix for 3 vulnerable dependencies (#395)
  [Snyk] Fix for 1 vulnerable dependencies (#394)
  Add text drop feature for seed input
  [Snyk] Fix for 1 vulnerable dependencies (#389)

# Conflicts:
#	src/desktop/src/libs/crypto.js
#	src/desktop/src/ui/views/onboarding/Login.js

Commit:
Fix unresolved conflict

Commit:
Merge branch 'develop' into feat/seed-storage-modules

# Conflicts:
#	src/shared/actions/accounts.js
#	src/shared/actions/wallet.js
#	src/shared/reducers/accounts.js

Commit:
Revert debugging changes

Commit:
- Update mobile SeedStore usage
- Use trits for address generation

Commit:
- Revert debug change
- Use `getIotaInstance` for `prepareTransfers`

rajivshah3 commented on src/mobile/src/libs/SeedStore/Keychain.js:
```diff
@@ -0,0 +1,155 @@
+            args = [...args, options];
+        }
+
+        return new Promise((resolve, reject) => {
```
It might be easier to just use `prepareTransfersAsync` from `shared/libs/iota/extendedApi`

rajivshah3 commented on src/mobile/src/libs/SeedStore/Keychain.js:
```diff
@@ -0,0 +1,155 @@
+    /**
+     * Get seed from keychain
+     * @param {boolean} rawTrits - Should return raw trits
+     * @returns {array} Derypted seed
```
`Derypted` -> `Decrypted`

rajivshah3 commented on src/mobile/src/libs/SeedStore/Keychain.js:
```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * Get seed from keychain
+     * @param {boolean} rawTrits - Should return raw trits
```
It doesn't look like this function takes a boolean parameter

Commit:
- Update code review
- Fix polling initialisation with no seedStore

Commit:
Merge branch 'develop' into feat/seed-storage-modules

# Conflicts:
#	src/shared/__tests__/libs/iota/addresses.spec.js
#	src/shared/libs/iota/addresses.js

laumair commented on src/desktop/src/ui/Index.js:
```diff
@@ -211,7 +212,10 @@ class App extends React.Component {
-                this.props.setOnboardingName('');
+                this.props.setAdditionalAccountInfo({
+                    additionalAccountName: '',
+                    addingAdditionalAccount: false,
```
Should we also reset `additionalAccountType` to default here?

laumair commented on src/desktop/src/ui/views/wallet/Index.js:
```diff
@@ -50,22 +50,30 @@ class Wallet extends React.PureComponent {
-            this.byteTritCheck();
-        }
+        //if (!this.props.completedByteTritSweep) {
+        this.byteTritCheck();
```
Are we allowing byte -> trit check on every mount?

laumair commented on src/desktop/src/ui/views/wallet/Index.js:
```diff
@@ -50,22 +50,30 @@ class Wallet extends React.PureComponent {
         });
 
         Promise.all(accounts).then((accounts) => {
+            accounts = accounts.filter((account) => account && account.accountName);
```
I don't think reassignment is allowed here with `accounts` being a `const`?

laumair commented on src/mobile/src/ui/views/onboarding/SetPassword.js:
```diff
@@ -116,58 +98,36 @@ class SetPassword extends Component {
+
+        const seedStore = new SeedStore.keychain(pwdHash);
+
+        const isUniqueueSeed = seedStore.isUniqueSeed(seed);
```
Typo. `isUniqueSeed`

laumair commented on src/shared/actions/polling.js:
```diff
@@ -431,7 +431,7 @@ export const getAccountInfo = (accountName, notificationFn) => {
         const existingAccountState = selectedAccountStateFactory(accountName)(getState());
 
-        return syncAccount()(existingAccountState, null, null, notificationFn)
+        return syncAccount()(existingAccountState, undefined, undefined, notificationFn)
```
`syncAccount` accepts three arguments, the second `undefined` is unnecessary.

Commit:
Code review bugfixes

Commit:
Remove unused dependency
